### DISCUSSION TITLE
HTML attributes are being stripped from menu item Navigation Label field, can I stop that?

### DISCUSSION THREAD

**Discussion by oxygensmith** [holger1411/understrap#1849] [Q&A] [answered]
Hi Understrap users and contributors,

I've tried searching / fiddling around with code in the class-wp-bootstrap-navwalker.php, but I'm unable to stop it from **stripping out class attributes from HTML tags I put in the Navigation Label field** in a WordPress menu item. 

I'd ideally use Font Awesome Pro 6 for tags for Navigation item labels, but I know we're only compatible up to FA5. I've tried Font Awesome Pro 5 and I can't quite seem to get it right. I know i can use Bootstrap icons, etc.

My use case is this: I typically put < span > tags in the Navigation Title bar to render text on mobile screens, and render icons on desktop/wider screens in a menu item. So the Navigation Label field value will look like this:

`<span class="fab fa-square-facebook nav-icon"></span><span class="nav-label">Facebook</span>`

and the SCSS will look something like

```
.nav-link .nav-icon {
    display: none;
}
.nav-link .nav-label {
    display: inherit;
}
@include media-breakpoint-up(md) {
  .nav-link .nav-icon {
      display: inherit;
  }
  .nav-link .nav-label {
      display: none;
  }
}
```

But when the menu is rendered, even with an unmodified Understrap theme (not my child theme), the Navigation Label is rendered as:

```
<a class="nav-link...">
  <span></span>
  <span>Facebook</span>
</a>

```

Understrap seems to strip all attributes and classnames, not just Font Awesome classes. So I can't replace it with say, Bootstrap icons, because I'll have the same problem.

It seems like there might be good reasoning behind stripping attributes in Navigation Labels (accessibility?). I'm open to doing this differently, but for my use case, it's most expedient to put markup right in the Navigation Label so I can change the icon that appears without coding specific icons in, say, through the stylesheet and pseudo-classes. But please let me know if there's a 'better-practice' way to accomplish what I want to do, if you know.

Thanks for anyone's help in advance!

**Comment by IanDelMar** [comment, marked as answer]
You could either remove the filter that strips the tags with
```php
remove_filter( 'the_title', 'understrap_kses_title' );
```
or adjust the allowed tag like so
```php
add_filter( 'understrap_kses_title', 'slug_allowed_tags_in_title' );
function slug_allowed_tags_in_title( $allowed_tags ) {
	$allowed_tags['span']['class'] = true;
	return $allowed_tags;
}
```
WordPress does not auto-escape titles. That's why the filter has been introduced.

**Reply by oxygensmith** [reply]
Hi Ian, thanks for replying. 

Is that filter supposed to affect the tags allowed in the **Navigation Label** field on a menu item in Wordpress' menu section? I am not trying to change the filter on a page or post title. 

What I am trying to do is detect where Understrap seems to be taking tags inserted into Navigation Label fields and stripping their attributes - perhaps somewhere in _class_wp_bootstrap_navwalker.php_. 

Values in Navigation Label fields override using the Page Title in your menu, for example, if your title is really really long and you need to display a value like "Issues." Or display icons, or whatever you want. Here's a screenshot of how I'm using it:

![span-classes-in-navigation-label-2](https://user-images.githubusercontent.com/51263/182671611-e90e66af-7e0b-484d-aebf-aa4aab1e1f8f.png)

(I use "Contact" in the Title attribute so only the text of the nav-item becomes the anchor link title attribute, not the rest of the HTML.)

The HTML in the Navigation Label field should display directly as the content inside of a.nav-link for that button. I haven't had this issue in Understrap before so I'm wondering if it's a recent update that changed this. 

To rule other possibilities than Understrap out, I did these tests: 
- I disabled plugins to make sure they weren't modifying/stripping Navigation Label attribute values. 
- I switched to Understrap theme temporarily to make sure my Child Theme wasn't doing anything.
- I switched to Wordpress Twenty Twenty theme temporarily and can verify that HTML in my Navigation Labels remains intact when I use the theme.

Thanks again

**Reply by IanDelMar** [reply]
I perfectly understood what you are trying to achieve. In case you have tried  the `add_filter` method, I haven't checked the shape of the `$allowed_tags` array and posted a wrong answer. I updated the code in my original answer.

> I am not trying to change the filter on a page or post title.

In fact you are. A menu item is a WP_Post object and the navigation label is what you get from `$menu_item->title`. If the item represents a post or a page this corresponds to `$post->post_title` where $post is the original object. Without a custom navigation label `$menu_item->title`is equal to `$post->post_title`.

The filter was introduced a while ago. I guess you haven't noticed yet because it is very common to use `<i>` tags for icons instead of `<span>`. For `<i>` the filter does not strip the class attribute.

**Reply by IanDelMar** [reply]
If you are interested, here you find the menu item setup https://github.com/WordPress/WordPress/blob/39903f841efad79a76fe8caa45957cae6948b6b5/wp-includes/nav-menu.php#L817-L995.

**Reply by oxygensmith** [reply]
OK, I tried your slugs_allowed_in_title filter/function again, and it worked! 

**Huge thanks.** I'm so glad that this is possible. You saved me much time poking around, Ian 🙌 

I also appreciate your insight/explanation that Navigation Label is part of $menu-Item->post_title (intuitively, it seems separate to me, which is probably why I went off looking in other directions for how the N.L. value was filtered or modified in some other way).

**Reply by IanDelMar** [reply]
As the issue is resolved you may want to mark it as answered. Then others know a) that you don't need support anymore and b) that they might find a solution here if they face the same problem.

I made a PR that address this issue. See #1850

**Reply by oxygensmith** [reply]
Thanks for making that PR, Ian. And, done. 
(Apologies, I couldn't find where to mark the discussion as answered, but I just figured out that a discussion thread has to be categorized as "Q&A" first. Fixed!)
